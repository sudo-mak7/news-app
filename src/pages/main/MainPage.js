import React, { useEffect } from 'react'
import { Container, Header, Loader } from 'semantic-ui-react'
import 'semantic-ui-css/semantic.min.css'
import { useDispatch, useSelector } from 'react-redux'
import {
  getNewsIdsErrorSelector,
  getNewsIdsLoaderSelector,
  getNewsIdsSelector,
  getNewsSelector,
  getNewsLoaderSelector,
  getNewsErrorSelector,
  getCurrentPageNumberSelector,
  getCurrentPageNewsSelector,
  getPagesLeftSelector
} from '@redux/selectors'
import { clearNewsIds, fetchNewsIds } from '@redux/news/newsIdsSlice'
import { clearNews, fetchNewsByIds } from '@redux/news/newsSlice'
import {
  setCurrentPageNumber,
  setCurrentPageNews,
  setPagesLeft,
  clearPaginationState
} from '@redux/pagination/paginationSlice'
import ButtonUpdateNewsComponent from '@components/ui/ButtonUpdateNewsComponent'
import MainPageNewsSectionComponent from '@components/news/MainPageNewsSectionComponent'
import { lazyLoading } from '@utils/lazyLoading'
import { windowScrollUp } from '@utils/windowScrollUp'

const MainPage = () => {
  const dispatch = useDispatch()

  const newsIds = useSelector(getNewsIdsSelector)
  const newsIdsIsLoading = useSelector(getNewsIdsLoaderSelector)
  const newsIdsLoadingError = useSelector(getNewsIdsErrorSelector)

  const news = useSelector(getNewsSelector)
  const newsIsLoading = useSelector(getNewsLoaderSelector)
  const newsLoadingError = useSelector(getNewsErrorSelector)

  const errorMessage = 'Error loading news :('

  const currentPage = useSelector(getCurrentPageNumberSelector)
  const currentPageNews = useSelector(getCurrentPageNewsSelector)
  const pagesLeft = useSelector(getPagesLeftSelector)

  useEffect(() => {
    windowScrollUp()
    if (!newsIds.length) {
      dispatch(fetchNewsIds())
    }
  }, [])

  useEffect(() => {
    if (news.length) {
      dispatch(setCurrentPageNews(news))
    }

    if (!pagesLeft && newsIds.length) {
      dispatch(setPagesLeft(newsIds.length - 1))
    }

    if (pagesLeft) {
      dispatch(setPagesLeft(pagesLeft - 1))
    }
  }, [news])

  useEffect(() => {
    if (!newsIdsIsLoading && !newsIsLoading) {
      return () => {

        dispatch(clearPaginationState())
        dispatch(clearNews())
        dispatch(clearNewsIds())
      }
    }
  }, [])

  useEffect(() => {
    if (currentPage !== 0 && !newsIdsIsLoading && currentPageNews.length) {
       dispatch(fetchNewsByIds(newsIds[currentPage]))
    }
  }, [currentPage])

  const lazyLoadingCallback = () => {
    dispatch(setCurrentPageNumber())
  }

  const target = document.querySelector('#loader')

  useEffect(() => {
    if (!newsIdsIsLoading && !newsIsLoading && target) {
      lazyLoading(lazyLoadingCallback, target)
    }
  }, [target])

  const loaderRender = <Loader active/>

  const errorRender =
    <Header as='h2' color='blue' textAlign='center'>
      { errorMessage }
    </Header>

  const newsRender = currentPageNews.map(n =>
    <MainPageNewsSectionComponent
      key={ n.id }
      { ...n }
    />
  )

  return (
    <main style={{ overflow: 'hidden' }}>
      <ButtonUpdateNewsComponent/>

      <Container text style={{ marginTop: '5em' }}>
        { newsIdsLoadingError
            ? errorRender
            : ( newsIdsIsLoading || newsIsLoading ) && !currentPageNews.length
                ? loaderRender
                : newsRender
        }

        { !newsIdsIsLoading
            && pagesLeft > 0
                ? <div id='loader' style={{ height: '50px' }}>
                    <Loader active inline='centered' />
                  </div>
                : newsLoadingError
                   ? errorRender
                   : ''
        }
      </Container>
    </main>
  )
}

export default MainPage
