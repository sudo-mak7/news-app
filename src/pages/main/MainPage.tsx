import * as React from 'react'
import { useEffect } from 'react'
import { Container, Header, Loader } from 'semantic-ui-react'
import 'semantic-ui-css/semantic.min.css'
import { useAppDispatch, useAppSelector } from '@redux/reduxHooks'
import {
  getNewsIdsErrorSelector,
  getNewsIdsLoaderSelector,
  getNewsIdsSelector,
  getNewsSelector,
  getNewsLoaderSelector,
  getNewsErrorSelector,
  getCurrentPageNumberSelector,
  getCurrentPageNewsSelector,
  getPagesLeftSelector
} from '@redux/selectors'
import { clearNewsIds, fetchNewsIds } from '@redux/news/newsIdsSlice'
import { clearNews, fetchNewsByIds } from '@redux/news/newsSlice'
import {
  setCurrentPageNumber,
  setCurrentPageNews,
  setPagesLeft,
  clearPaginationState
} from '@redux/pagination/paginationSlice'
import ButtonUpdateNewsComponent from '@components/ui/ButtonUpdateNewsComponent'
import MainPageNewsSectionComponent from '@components/news/MainPageNewsSectionComponent'
import { windowScrollUp } from '@utils/windowScrollUp'
import { NewsInterface } from '@common/types/newsInterface'
import { ERROR_LOADING_NEWS_MESSAGE } from '@common/messages/errors'
import { useInView } from 'react-intersection-observer'

const MainPage = (): JSX.Element => {
  const dispatch = useAppDispatch()

  const newsIds = useAppSelector(getNewsIdsSelector)
  const newsIdsIsLoading = useAppSelector(getNewsIdsLoaderSelector)
  const newsIdsLoadingError = useAppSelector(getNewsIdsErrorSelector)

  const news = useAppSelector(getNewsSelector)
  const newsIsLoading = useAppSelector(getNewsLoaderSelector)
  const newsLoadingError = useAppSelector(getNewsErrorSelector)

  const currentPage = useAppSelector(getCurrentPageNumberSelector)
  const currentPageNews = useAppSelector(getCurrentPageNewsSelector)
  const pagesLeft = useAppSelector(getPagesLeftSelector)

  const { ref, inView } = useInView({
    threshold: 0
  })

  const conditionToFetchNewsByIds = () => {
    return currentPage !== 0 && !newsIdsIsLoading && currentPageNews.length
  }

  const conditionToDoLazyLoading = () => {
    return !newsIdsIsLoading && !newsIsLoading && inView
  }

  const conditionToNewsRender = () => {
    return  ( newsIdsIsLoading || newsIsLoading ) && !currentPageNews.length
  }

  useEffect(() => {
    windowScrollUp()
    if (!newsIds.length) {
      dispatch(fetchNewsIds())
    }

    if (!newsIdsIsLoading && !newsIsLoading) {
      return () => {
        dispatch(clearPaginationState())
        dispatch(clearNews())
        dispatch(clearNewsIds())
      }
    }
  }, [])

  useEffect(() => {
    if (news.length) {
      dispatch(setCurrentPageNews(news))
    }

    if (!pagesLeft && newsIds.length) {
      dispatch(setPagesLeft(newsIds.length - 1))
    }

    if (pagesLeft) {
      dispatch(setPagesLeft(pagesLeft - 1))
    }
  }, [news])

  useEffect(() => {
    if (conditionToFetchNewsByIds()) {
       dispatch(fetchNewsByIds(newsIds[currentPage]))
    }
  }, [currentPage])

  useEffect(() => {
    if (conditionToDoLazyLoading()) {
      dispatch(setCurrentPageNumber())
    }
  }, [inView])

  const errorRender =
    newsLoadingError
      ? <Header
          as='h2'
          color='blue'
          textAlign='center'>
          { ERROR_LOADING_NEWS_MESSAGE }
        </Header>
      : ''

  const newsRender =
    conditionToNewsRender()
        ? <Loader active/>
        : currentPageNews.map((n: NewsInterface) =>
            <MainPageNewsSectionComponent
              key={ n.id }
              { ...n }
            />
    )

  return (
    <main style={{ overflow: 'hidden' }}>
      <ButtonUpdateNewsComponent/>

      <Container text style={{ marginTop: '5em' }}>
        { newsIdsLoadingError
            ? errorRender
            : newsRender
        }

        { !newsIdsIsLoading && pagesLeft > 0
            ? <div
                ref={ ref }
                style={{ height: '50px' }}
              >
                <Loader active inline='centered' />
              </div>
            : errorRender
        }
      </Container>
    </main>
  )
}

export default MainPage
