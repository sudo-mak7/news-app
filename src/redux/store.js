import { legacy_createStore as createStore, compose, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './reducers/rootReducer'
import { newsWatcher } from './sagas/rootSaga'

const sagaMiddleware = createSagaMiddleware()

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose

const configureStore = (preloadedState) => createStore(
  rootReducer,
  preloadedState,
  composeEnhancers(applyMiddleware(sagaMiddleware))
)

const store = configureStore({})

sagaMiddleware.run(newsWatcher)

export default store
