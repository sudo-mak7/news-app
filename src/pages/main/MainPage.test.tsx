import * as React from 'react'
import * as reduxHooks from '@redux/reduxHooks'
import { screen, render } from '@testing-library/react'
import { store } from '@redux/store'
import { Provider } from 'react-redux'
import MainPage from '@pages/main/MainPage'

const mainPage = (): JSX.Element =>
  <Provider store = { store } >
    <MainPage />
  </Provider>

const mockedDispatch = jest.spyOn(reduxHooks, 'useAppDispatch')

describe('MainPage test', () => {
  test('MainPage should to match snapshot', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(mainPage())
    expect(mainPage()).toMatchSnapshot()
  })

  test('renders MainPage ButtonUpdateNewsComponent', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(mainPage())

    const buttonInMainPage = screen.getByRole('button')
    expect(buttonInMainPage).toHaveClass('ui blue big icon button')
  })

  test('renders MainPage container', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(mainPage())

    const buttonInMainPage = screen.getByTestId('container')
    expect(buttonInMainPage).toBeInTheDocument()
  })

  test('dispatch should called once', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(mainPage())

    expect(dispatch).toHaveBeenCalledTimes(1)
  })

  afterAll(() => {
    jest.clearAllMocks()
  })
})