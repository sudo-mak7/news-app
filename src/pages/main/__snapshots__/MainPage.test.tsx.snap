// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage test MainPage should to match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPage />
</Provider>
`;

exports[`MainPage test renders MainPage ButtonUpdateNewsComponent 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPage />
</Provider>
`;

exports[`MainPage test renders MainPage container 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPage />
</Provider>
`;

exports[`MainPage test renders MainPage loader 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPage />
</Provider>
`;
